rsocket/RSocketResponder.cpp

```diff
@@ -70,35 +70,51 @@ RSocketResponder::handleRequestChannelCore(
     }
 
     void onError(folly::exception_wrapper ex) noexcept override {
-      DCHECK(inner_);
-      inner_->onError(std::move(ex));
+      VLOG(3) << "handleRequestChannelCore::onError: " << ex.what();
+      std::lock_guard<std::mutex> guard(errorMutex_);
+      if (inner_) {
+        inner_->onError(std::move(ex));
 
-      inner_.reset();
-      subscription_.reset();
+        inner_.reset();
+        subscription_.reset();
+        error_.reset();
+      } else {
+        error_ = std::move(ex);
+      }
     }
 
     void subscribe(
         yarpl::Reference<yarpl::flowable::Subscriber<rsocket::Payload>> inner) {
       CHECK(!inner_); // only one call to subscribe is supported
       CHECK(inner);
+      std::lock_guard<std::mutex> guard(errorMutex_);
+
       inner_ = std::move(inner);
       if (subscription_) {
         inner_->onSubscribe(subscription_);
       }
+
+      // it's possible to get an error before subscribe happens, delay sending
+      // it but send it when this class gets subscribed
+      if (error_) {
```

Comment:
You can move this inside of the `if (subscription_)` check, and also do `onError(std::move(error_))`.